### PR TITLE
MDS-1599 Feature/ Add MA & HSRC tables, Support Variances in the Database

### PR DESCRIPTION
**Pull request by NWCalvank**
This includes the MDS-1551 (HSRC code table) database changes as well as most of the required database changes for MDS-1599 (variances). The remaining work to be done for variances is surrounding document management.

These changes will likely be pushed to this PR before merging, but I wanted to open discussion on the current changes as soon as possible. Based on consensus around these changes, we can add the database support for associated documents.

### CODE REVIEW

File: migrations/sql/V2019.03.25.11.06__add_hsrc_code.sql [hunk 1]
```diff
@@ -0,0 +1,1069 @@
...
+    article_act_code character varying(5)     PRIMARY KEY                        ,
+    description      character varying(100)                              NOT NULL,
+    effective_date   date                     DEFAULT now()              NOT NULL,
+    expiry_date      date                     DEFAULT '9999-12-31'::date NOT NULL,
+    update_timestamp timestamp with time zone DEFAULT now()              NOT NULL
+);
+
+
+-- Comments
+COMMENT ON TABLE article_act_code IS 'Contains a list of values used to indicate an Act or Regulation and is used to define the article_act_code. I.e. MA - Mines Act, HSRC - Health, Safety and Reclamation Code for Mines in BC.';
```
Comment (klefler) [inline migrations/sql/V2019.03.25.11.06__add_hsrc_code.sql]: I think I made a mistake when providing an example of a comment for the article_act_code table. The last bit doesn't make as much sense today... Can you please update it?

"Contains a list of values used to indicate an Act or Regulation. I.e. MA - Mines Act, HSRCM - Health, Safety and Reclamation Code for Mines in BC."

File: migrations/sql/V2019.03.25.11.06__add_hsrc_code.sql [hunk 2]
```diff
@@ -0,0 +1,1069 @@
...
+    compliance_article_id integer                  PRIMARY KEY                        ,
+    article_act_code      character varying(5)                                NOT NULL,
+    section               character varying(6)                                NOT NULL,
+    sub_section           character varying(4)                                        ,
+    paragraph             character varying(4)                                        ,
+    sub_paragraph         character varying(4)                                        ,
+    description           character varying(100)                              NOT NULL,
+    effective_date        date                     DEFAULT now()              NOT NULL,
+    expiry_date           date                     DEFAULT '9999-12-31'::date NOT NULL
+);
```
Comment (klefler) [inline migrations/sql/V2019.03.25.11.06__add_hsrc_code.sql]: I forgot to mention that we may want the 4 audit columns on this table (create_user, create_timestamp, update_user, update_timestamp)

File: migrations/sql/V2019.03.25.14.40__add_variance_table.sql [hunk 3]
```diff
@@ -0,0 +1,33 @@
+CREATE TABLE IF NOT EXISTS variance
+(
+    variance_id           integer                PRIMARY KEY                           ,
+    compliance_article_id integer                                              NOT NULL,
+    mine_guid             uuid                                                 NOT NULL,
+    note                  character varying(300) DEFAULT ''::character varying NOT NULL,
+    issue_date            date                   DEFAULT '9999-12-31'::date    NOT NULL,
+    received_date         date                   DEFAULT '9999-12-31'::date    NOT NULL,
+    expiry_date           date                   DEFAULT '9999-12-31'::date    NOT NULL
+);
```
Comment (klefler) [inline migrations/sql/V2019.03.25.14.40__add_variance_table.sql]: We probably should have the 4 audit columns on this table too.

Comment (NWCalvank) [reply migrations/sql/V2019.03.25.14.40__add_variance_table.sql]: Added 👍

File: migrations/sql/V2019.03.25.11.06__add_hsrc_code.sql (hunk 2 above)
Comment (NWCalvank) [reply migrations/sql/V2019.03.25.11.06__add_hsrc_code.sql]: Added 👍

**Comment by NWCalvank** [comment]
Closing to do a full clean-up 🧹

**Comment by NWCalvank** [comment]
Reopening. Let's hope for a beautiful and successful build

**Comment by NWCalvank** [comment]
Everything is in a broken state. Going to include a very manual cleanup

File: migrations/sql/V2019.03.25.14.40__add_variance_table.sql [hunk 4]
```diff
@@ -35,3 +35,17 @@ ALTER TABLE ONLY variance
...
+    variance_document_xref_guid uuid    DEFAULT gen_random_uuid() NOT NULL,
+    mine_document_guid          uuid                              NOT NULL,
+    variance_id                 integer                           NOT NULL
+);
+
+-- Constraints
+ALTER TABLE ONLY variance_document_xref
+    ADD CONSTRAINT variance_document_xref_mine_document_guid_fkey FOREIGN KEY (mine_document_guid) REFERENCES mine_document(mine_document_guid);
+ALTER TABLE ONLY variance_document_xref
+    ADD CONSTRAINT variance_document_xref_variance_id_fkey FOREIGN KEY (variance_id) REFERENCES variance(variance_id);
```
Comment (klefler) [inline migrations/sql/V2019.03.25.14.40__add_variance_table.sql]: Looks good to me! Thanks.

**Comment by NWCalvank** [comment]
Reopening now that resources have been addressed. This should fix build issues 🤞

**Comment by NWCalvank** [comment]
Attempting another cleanup -- but with a more meticulous, manual effort.

**Comment by NWCalvank** [comment]
Third time's a charm ⭐️

**Pull request merged**